fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,544 @@
+            should_include_retired_keys: true,
+            // If a key has expired- i.e. its last scanned block is greater
+            // than or equal to its public expiry- then it will no longer scan
+            // blocks. Therefore, we need to include unexpired keys because they
+            // are still supposed to be scanned by Fog.
+            should_only_include_unexpired_keys: true,
+        };
+
+        // First, find the "inactive_outstanding_keys" which are outstanding
+        // keys that we've grabbed from the RecoveryDb.
+        //
+        // TODO: Add a config that allows us to set this start block.
+        let ingress_public_key_records: Vec<IngressPublicKeyRecord> =
+            self.recovery_db.get_ingress_key_records(
+                /* start_block_at_least= */ 0,
+                outstanding_keys_filters,
+            )?;
+
+        Ok(ingress_public_key_records
+            .iter()
+            .map(|record| record.key)
+            .collect())
+    }
+
+    /// Tries to activate a node with the given inactive outstanding key.
+    fn try_to_activate_node_for_inactive_outstanding_key(
+        &self,
+        inactive_outstanding_key: &CompressedRistrettoPublic,
+        ingest_summary_node_mappings: &[IngestSummaryNodeMapping],
+    ) -> Result<(), OverseerError> {
+        log::debug!(
+            self.logger,
+            "Trying to activate an idle node with inactive outstanding key: {:?}",
+            &inactive_outstanding_key
+        );
+        for ingest_summary_node_mapping in ingest_summary_node_mappings {
+            let node_ingress_key = match CompressedRistrettoPublic::try_from(
+                ingest_summary_node_mapping
+                    .ingest_summary
+                    .get_ingress_pubkey(),
+            ) {
+                Ok(key) => key,
+                Err(_) => continue,
+            };
+            if inactive_outstanding_key.eq(&node_ingress_key) {
+                // We have to keep trying to activate this node with the
+                // matching ingress key until it becomes active.
+                //
+                // This prevents the secenario in which the active node is
+                // bounced, and then we end up here and we're trying to
+                // activate an idle node. If we give up after a few
+                // unsuccessful tries, then we'd report the previously
+                // active key as lost. But then the previously active server
+                // could be restarted by Kubernetes, have this lost key, and
+                // restart in the active state. This would be bad because
+                // we don't want a server to be active with a lost key.
+                //
+                // Since we always expect the node to come back in some
+                // time, we should just keep trying here.
+                //
+                // TODO: Add alerting if this is taking too long.
+                loop {
```

Comment:
I think it would be better not to do this in a loop, we could just return the error to the caller?

Because the information we used to decide to activate this node could become stale -- what if the operator is also doing something in background for example. If it is in this loop for 20 minutes then the overseer might be attempting to do something that no longer makes sense

Reply:
Agreed. As mentioned in an above comment, I moved this logic to the `perform_automatic_failover `method, and removed this loop. A failure to activate the node results in an error being propagated to the caller and no activation logic proceeds.